File: jianzhioffer/firstround/jianzhi_34.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        节点值和sum值可能为正也可能为负
        """
        def recur(node, subsum):
            if not node:
                return
            subsum += node.val
            path.append(node.val)
            if subsum == sum and not node.left and not node.right:
                # 记录路径时若直接执行res.append(path)，则是将path对象加入了res后续path改变时，
                # res中的path对象也会随之改变。正确做法：res.append(list(path))，
                # 相当于复制了一个 path 并加入到 res 。
                res.append(list(path))
            if node.left:
                recur(node.left, subsum)
            if node.right:
                recur(node.right, subsum)
            path.pop()

        if not root:
            return []
        res, path = [], []
        recur(root, 0)
        return res


if __name__ == '__main__':
    l1 = TreeNode(1)
    l2 = TreeNode(0)
    l3 = TreeNode(1)
    l4 = TreeNode(1)
    l5 = TreeNode(2)
    l6 = TreeNode(0)
    l7 = TreeNode(-1)
    l8 = TreeNode(0)
    l9 = TreeNode(1)
    l10 = TreeNode(-1)
    l11 = TreeNode(0)
    l12 = TreeNode(-1)
    l13 = TreeNode(0)
    l14 = TreeNode(1)
    l15 = TreeNode(0)

    l1.left = l2
    l1.right = l3
    l2.left = l4
    l2.right = l5
    l3.left = l6
    l3.right = l7
    l4.left = l8
    l4.right = l9
    l5.left = l10
    l5.right = l11
    l6.left = l12
    l6.right = l13
    l7.left = l14
    l7.right = l15

    obj = Solution()
    result = obj.pathSum(l1, 2)
    print(result)

File: jianzhioffer/secondround/jianzhi_34.cpp
 #include <vector>
 #include <cmath>
 using namespace std;

 //Definition for a binary tree node.
 struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };

//前序遍历的回溯法。想到了前序遍历+递归，但回溯这里的代码实现没有写出来。
//时间复杂度O(N)
//空间复杂度O(N)。最坏情况下二叉树退化为链表，path数组在遍历过程中可能需要存储链表的所有节点，占用O(N)空间
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        if(root==NULL) return res;
        obj_num = sum;
        dfs(root, 0);
        return res;
    }

    void dfs(TreeNode* root, int cur)
    {
        if(root==NULL) return;
        cur += root->val;
        path.push_back(root->val);

        if(cur==obj_num && root->left==NULL && root->right==NULL)
        {
            res.push_back(path);
            path.pop_back();
            return;
        }
        dfs(root->left, cur);
        dfs(root->right, cur);
        path.pop_back();
        return ;
    }

private:
    int obj_num;
    vector<vector<int>> res;
    vector<int> path;
};

int main(int argc, char const *argv[])
{
    TreeNode t1 = TreeNode(5);
    TreeNode t2 = TreeNode(4);
    TreeNode t3 = TreeNode(8);
    TreeNode t4 = TreeNode(11);
    //TreeNode t5 = TreeNode(5);
    TreeNode t6 = TreeNode(13);
    TreeNode t7 = TreeNode(4);
    TreeNode t8 = TreeNode(7);
    TreeNode t9 = TreeNode(2);
    TreeNode t10 = TreeNode(5);
    TreeNode t11 = TreeNode(1);

    t1.left = &t2;
    t1.right = &t3;
    t2.left = &t4;
    // t2.right = &t5;
    t3.left = &t6;
    t3.right = &t7;
    t4.left = &t8;
    t4.right = &t9;
    t7.left = &t10;
    t7.right = &t11;

    Solution obj;
    vector<vector<int>> result = obj.pathSum(&t1, 22);
    return 0;
}